lte/gateway/c/oai/tasks/mme_app/mme_app_edns_emulation.c

```diff
@@ -60,6 +60,7 @@ int mme_app_edns_add_sgw_entry(bstring id, struct in_addr in_addr) {
       if (HASH_TABLE_OK == rc) return RETURNok;
     }
   }
+  free(cid);
```

Comment:
Awesome! Finding real memory leaks and fixing them!

## Minor Change Request

My recommendation on this would be to move the free() up to just before line 60. The rationale is that when others come along and modify this function, they could add additional return statements that short-circuit your free (not seeing it).  I think it's best to free the resource as soon as you are done with it.

## Better Yet - But not this PR? File a GH Issue for discussion?

We might consider a broader campaign to improve the safety of our allocations like this one (where the resource should clearly be freed when context exits).  A better solution in this type of context would even be to consider a RAII style approach in C.  It appears GCC (and maybe Clang but undocumented) support the `__cleanup__` variable attribute, which would look something like this.

```c
// Some shared utility header
void clean_up_char_array(char* to_be_freed)
{
  free(to_be_freed);
}

char* cid __attribute__ ((__cleanup__(clean_up_char_array))) = calloc(1, blength(id) + 1) ;
if (stuff){
}
// No matter what happens, when cid goes out of context we call clean_up_char_array(cid).
```

Reply:
I'll definitely apply the change you mentioned, but I think it makes sense to move the RAII point into a larger discussion because I feel unsure about the implications associated with the compiler specificity of using GCC. Should I file the GitHub issue associated with this - if you have broader vision and context and want to do it, go for it, but I also can.

Reply:
I spoke with some of the team about the `__cleanup__` RAII option - and @amarpad brought up the excellent point that we are going to be supporting ARM more and more. Looking into it, I [could not find support for __cleanup__ by the armcc compiler](https://developer.arm.com/documentation/dui0472/k/Compiler-specific-Features/Function-attributes?lang=en).  Further, I do see `__cleanup__` used in linux's `systemd`, which surprises me (how can this not support ARM?).  But perhaps that is a [rogue linux agent](https://lwn.net/Articles/589433/).  Sadly, I think we have to avoid __cleanup__ use.  Another option I'm finding is that some have authored smart-pointer like libraries for c. But that's another discussion =D

Reply:
As to testing this PR, I realize you have Vagrant issues that are blocking you. There was a regression on our vagrant based E2E testing this week - or a couple regressions. Rebasing to master might fix your issue?  If we cannot get this working for you in reasonable time, we could rely on the OAI CI test coverage which seems reasonable to me for this PR (and not run local Vagrant tests).

Reply:
Vagrant VM is still broken due to migration to Ubuntu 20.04 effort - you should omit that test and rely entirely on CI tests for this PR - you have my OK.

Reply:
Sounds good, I'll make the change later this morning. Is there anything I need to do to omit the local vagrant tests within the context of this PR?